### PR TITLE
feat(challenge): Endpoint for Related Challenges (#563)

### PR DESCRIPTION
**Pull request by isaac-diez**
### Summary

This PR introduces the GET /challenges/{challengeId}/related endpoint in the ChallengeController, which retrieves up to three challenges related to the given challengeId. The relation is determined based on shared attributes such as language, difficulty level, and optionally tags.

The controller delegates the logic to the ChallengeService, which fetches the current challenge and then filters all others by the defined criteria. The related challenges are returned in a GenericResultDto wrapper with pagination info. A random subset of up to three challenges is selected to ensure variation across multiple requests.

Integration tests were added using WebTestClient to validate the endpoint in different scenarios. Manual testing was also performed with Postman to ensure the expected behavior in local environments.

### Changes
**ChallengeController.java**

-  Added GET /challenges/{challengeId}/related endpoint to retrieve up to 3 related challenges.
- Annotated with OpenAPI documentation (@Operation, @ApiResponse).
- Returns:
  - 200 OK with a wrapped list of related challenges
  - 204 No related challenges found
  - 400 Bad Request for Malformed, missing or invalid parameters
  - 404 The Challenge with given Id was not found

**ChallengeService.java**
- Implemented logic to:
  - Fetch the base challenge by UUID
  - Filter other challenges based on shared language, difficulty, and tags with buildFilterPredicate() in order to avoid code repetition from method getChallengesByFilter() (which should be refactored to use it as well)
  - Shuffle results and return up to 3 random items
  - Returns a GenericResultDto<ChallengeDto> via a Mono
  - Ensures that the base challenge is excluded from the result

**ChallengeControllerTest.java**
- Added test for case where 3 related challenges are returned - code 200
- Added test for case where no related challenges are found - code 204
- Added test for case when main challenge is not found  - code 404
- Added test for case when not a valid UUID is passed - code 400
- All tests use WebTestClient and mock the service with Mockito

**ChallengeServiceTest.java**
- Added unit tests for:
  - No related challenges → empty result
  - Fewer than 3 matching challenges → all returned
  - More than 3 matching challenges → random 3 selected
  - When UUID is invalid → BadUUIDException should be thrown
  - When the language id is empty → shouldn't filter by language
  - When the language id is null → should be excluded
  - When level is empty → shouldn't filter by level
  - When tags exist but challenge has no tags → should be excluded
- All dependencies are mocked
- Assertions validate that results respect filtering and count constraints

in Postman, confirmed that:
- When 3 related challenges exist, they are returned as expected.
- When no matches exist (e.g. due to non-overlapping tags), an empty array is returned with code 204.
- Responses vary on repeated requests due to the randomization of related challenges.

**CHANGELOG.md**
- Added corresponding update:
  - - Added GET endpoint in `ChallengeController` to retrieve related challenges to the one on screen of same language, difficulty and tag (Taiga [#563], PR [#947])

**.env.CI.dev**
- added corresponding update:
  - - MICROSERVICE_VERSION_itachallenge-challenge=2.4.0-RELEASE 

## **TECHNICAL DEBT**
Two issues have been registered in Taiga to resolve in future PR:
- [Issue 579](https://tree.taiga.io/project/luisvi-ita-challenges/issue/579) [BE] **Refactor getChallengesByFilter()**:
Analyze and standardize, as far as possible, the functionalities of the filter search and the search for associated challenges. getChallengesByFilter() and getRelatedChallenges() share the same filters so it is suggested to reuse buildFilterPredicate() and thus comply with the DRY principle.
- [Issue 580](https://tree.taiga.io/project/luisvi-ita-challenges/issue/580) [BE] **[BE] Apply a single criterion (200 + empty list vs. 204)**:
Define and apply a single criterion (200 + empty list vs. 204) to all lists and notify the frontend.

### Manual testing
Manually verified the endpoint with Postman using valid challengeId values.

## 📋 PR Author Self-check
✅ PR is focused and isolated
✅ Branch follows naming conventions: feature/563-related-challenges.
✅ All logic tested locally.
✅ No debug logs left.
✅ Exception mapped correctly.
✅ Tests included and passing.
✅ PR clearly documents scope and purpose.

### CODE REVIEW

**Review by acocinas — APPROVED** [review]
✅ Pull Request Review – #563: Related Challenges Endpoint
Thanks for the clear and well-documented PR.

I’ve reviewed both the code and the functionality, and I want to highlight the following:

💡 General Feedback
I really appreciated the live demo you provided to showcase how the /related endpoint works and prove that it behaves as expected. That gave me full confidence in the logic behind it.

The implementation is clean, focused, and fully aligned with the scope described in the PR summary.

Code is well-structured and readable, with proper delegation to the service layer and clear separation of concerns.

🔍 Code Quality
The filtering logic in ChallengeServiceImpl is robust, handling:

Empty or missing language, level or tags gracefully.

Random selection of up to 3 results to ensure variability.

Use of Optional, Mono, and Collections.shuffle is appropriate and idiomatic for the problem.

The controller is well annotated with OpenAPI, and response scenarios are clearly handled (200, 400).

The GenericResultDto wrapping makes the response structure consistent across the app.

🧪 Tests
Unit and integration tests cover:

All edge cases (no matches, less than 3, more than 3).

Filtering logic with missing or null fields.

Randomization of the result list.

Test coverage is excellent and ensures future maintainability.

🧪 Manual testing
I’ve manually tested the endpoint locally using Postman and also verified:

Correct response for valid challengeId.

Empty array when no matches found.

400 error on malformed UUID.

Random variation on repeated requests.

🟢 Conclusion
Everything works as expected both in the code and in runtime. Thanks again for the demo and the solid work!

✅ Approved. Great job!

**Comment by ivilarop** [comment]
@isaac-diez @acocinas The PR name does not follow the convention indicated in the PR checklist

File: itachallenge-challenge/src/main/java/com/itachallenge/challenge/controller/ChallengeController.java [hunk 1]
```diff
@@ -165,6 +165,23 @@ public Flux<GenericResultDto<ChallengeDto>> getChallengesByFilter(@ModelAttribut
...
+            responses = {
+                    @ApiResponse(responseCode = "200", content = {@Content(schema = @Schema(implementation = ChallengeDto.class), mediaType = "application/json")}),
+                    @ApiResponse(responseCode = "200", description = "The language with given Id was not found."),
+                    @ApiResponse(responseCode = "400", description = "Missing or unexpected parameters"),
+                    @ApiResponse(responseCode = "400", description = "Malformed UUID")
+            })
+
+    public Mono<GenericResultDto<ChallengeDto>> getRelatedChallenges(@PathVariable("challengeId") String challengeId) {
+
+        return challengeService.getRelatedChallenges(challengeId);
```
Comment (ivilarop) [inline itachallenge-challenge/src/main/java/com/itachallenge/challenge/controller/ChallengeController.java]: I've noticed that the getRelatedChallenges method returns a Mono<GenericResultDto<ChallengeDto>> without encapsulating it in a ResponseEntity. I recommend checking if this approach adequately covers all scenarios from the perspective of the HTTP return codes (e.g., 200, 204, 400, etc.).

👁‍🗨 It may be helpful to compare how status codes are being returned on other endpoints in the project, especially when no results are found or there are input errors. This way, you can confirm if this endpoint is aligned with the expected behavior.

File: itachallenge-challenge/src/main/java/com/itachallenge/challenge/controller/ChallengeController.java [hunk 2]
```diff
@@ -165,6 +165,23 @@ public Flux<GenericResultDto<ChallengeDto>> getChallengesByFilter(@ModelAttribut
...
     }
 
+    @GetMapping("/challenges/{challengeId}/related")
+    @Operation(
+            operationId = "Get 3 related challenges on the \"Relacionat\" tab of a Challenge detail by (language, difficulty, or tags).",
+            summary = "Get to see 3 related challenges on a challenge detail page and their levels, details and their depending on the first language, difficulty and any tags.",
+            description = "Requesting 3 related challenges for the challenge the user is actually looking at or working on.",
+            responses = {
+                    @ApiResponse(responseCode = "200", content = {@Content(schema = @Schema(implementation = ChallengeDto.class), mediaType = "application/json")}),
+                    @ApiResponse(responseCode = "200", description = "The language with given Id was not found."),
```
Comment (ivilarop) [inline itachallenge-challenge/src/main/java/com/itachallenge/challenge/controller/ChallengeController.java]: Does it make sense to return 200 in that case?

File: itachallenge-challenge/src/main/java/com/itachallenge/challenge/controller/ChallengeController.java [hunk 3]
```diff
@@ -165,6 +165,23 @@ public Flux<GenericResultDto<ChallengeDto>> getChallengesByFilter(@ModelAttribut
...
+                    @ApiResponse(responseCode = "400", description = "Missing or unexpected parameters"),
```
Comment (ivilarop) [inline itachallenge-challenge/src/main/java/com/itachallenge/challenge/controller/ChallengeController.java]: There are two responses with responseCode = "400" but with different descriptions:

"Missing or unexpected parameters"

"Malformed UUID"
👉 Would it be helpful to unify or better distinguish them with more specific codes?

File: itachallenge-challenge/src/main/java/com/itachallenge/challenge/service/ChallengeServiceImpl.java [hunk 4]
```diff
@@ -117,6 +117,46 @@ public Flux<GenericResultDto<ChallengeDto>> getChallengesByFilter(
                 });
     }
 
+    @Override
+    public Mono<GenericResultDto<ChallengeDto>> getRelatedChallenges(String challengeId) {
```
Comment (ivilarop) [inline itachallenge-challenge/src/main/java/com/itachallenge/challenge/service/ChallengeServiceImpl.java:122]: The getRelatedChallenges method concentrates a lot of transformation, filtering, and conversion logic into a single block.

in any case when reviewing this method, we notice that it reuses some of the logic already implemented in getChallengesByFilter(), but it isn't directly leveraging it.

It might be interesting to review whether there is any opportunity to reuse or unify filters already implemented there to avoid duplication of logic.

Even if the entire method can't be reused due to the specifics of this endpoint, extracting shared logic into private methods could facilitate future reuse and improve readability.

Remember that if you reuse existing code, it must be specified in the PR description.

File: itachallenge-challenge/src/main/java/com/itachallenge/challenge/service/ChallengeServiceImpl.java [hunk 5]
```diff
@@ -117,6 +117,46 @@ public Flux<GenericResultDto<ChallengeDto>> getChallengesByFilter(
...
+                                Collections.shuffle(challengeDtos);
+                                List<ChallengeDto> selectedChallenges = challengeDtos.stream()
+                                        .limit(3)
+                                        .toList();
+                                GenericResultDto<ChallengeDto> resultDto = new GenericResultDto<>();
+                                resultDto.setInfo(0, 3, selectedChallenges.size(), selectedChallenges.toArray(new ChallengeDto[0]));
+                                return resultDto;
+                            });
+                })
+                .switchIfEmpty(Mono.just(new GenericResultDto<>()));
```
Comment (ivilarop) [inline itachallenge-challenge/src/main/java/com/itachallenge/challenge/service/ChallengeServiceImpl.java]: I see .switchIfEmpty(Mono.just(new GenericResultDto<>())) being used to return an empty result when no related challenges are found. Is the frontend team aware that an empty GenericResultDto in this context should be interpreted as "no related challenges" and not as an error?

👉 It might be helpful to align this behavior with other endpoints that return listings, or confirm that this case is handled correctly in the UI (a "No related challenges found" message, for example).

**Review by ivilarop — CHANGES_REQUESTED** [review]

File: itachallenge-challenge/src/main/java/com/itachallenge/challenge/controller/ChallengeController.java (hunk 2 above)
Comment (isaac-diez) [reply itachallenge-challenge/src/main/java/com/itachallenge/challenge/controller/ChallengeController.java]: You are right. I re-used the response codes in /challenges/byFilter without question. I've removed it.

File: itachallenge-challenge/src/main/java/com/itachallenge/challenge/controller/ChallengeController.java (hunk 3 above)
Comment (isaac-diez) [reply itachallenge-challenge/src/main/java/com/itachallenge/challenge/controller/ChallengeController.java]: I've improved the whole response codes situation, removing code 200 and adding a 204 when no related challenge is found, a 404 when the (man) challenge is not found and unified the two 400 into one.
I had to change the mock behavior for the test to get me a Mono.error(...) when invalid UUID since it gets a null by default.

File: itachallenge-challenge/src/main/java/com/itachallenge/challenge/controller/ChallengeController.java (hunk 1 above)
Comment (isaac-diez) [reply itachallenge-challenge/src/main/java/com/itachallenge/challenge/controller/ChallengeController.java]: No, it didn't cover it. I've fixed the code encapsulating it in a ResponseEntity as you pointed out.

File: itachallenge-challenge/src/main/java/com/itachallenge/challenge/service/ChallengeServiceImpl.java (hunk 5 above)
Comment (isaac-diez) [reply itachallenge-challenge/src/main/java/com/itachallenge/challenge/service/ChallengeServiceImpl.java]: As far as I know, no, the frontend doesn't know about this. But now the code doesn't return an empty GenericResultDto, it returns a 204 with "No related challenges found" message.
FYI no other endpoint is returning a 204 as of today.

File: itachallenge-challenge/src/main/java/com/itachallenge/challenge/service/ChallengeServiceImpl.java (hunk 4 above)
Comment (isaac-diez) [reply itachallenge-challenge/src/main/java/com/itachallenge/challenge/service/ChallengeServiceImpl.java:122]: I've refactored the method in the service adding a buildFilterPredicate() that acts as filter. It would be a good idea to refactor the method getChallengesByFilter() to use the new method and avoid the Don't Repeat Yourself principle.

**Review by acocinas — CHANGES_REQUESTED** [review]
Sorry buddy, but I think that while reviewing a PR, you may have unintentionally added some classes from that microservice to this PR as if they were your own. Please check the classes you've uploaded from the User microservice — I believe some of them were added to this PR by mistake during the review of another PR.

**Review by acocinas — APPROVED** [review]
Now I think it's all right

File: itachallenge-challenge/src/main/java/com/itachallenge/challenge/service/ChallengeServiceImpl.java (hunk 4 above)
Comment (ivilarop) [reply itachallenge-challenge/src/main/java/com/itachallenge/challenge/service/ChallengeServiceImpl.java:122]: Good job extracting buildFilterPredicate(). We won't be touching getChallengesByFilter() in this task because it's not the goal, but we must create the technical debt:

Refactor getChallengesByFilter() to reuse buildFilterPredicate() and thus comply with the DRY principle.

Please create a Taiga card (Tech Debt type) with this description. This way, it will be recorded and prioritized for the next sprint.

Comment (ivilarop) [reply itachallenge-challenge/src/main/java/com/itachallenge/challenge/service/ChallengeServiceImpl.java:122]: Other point:
Does it have more than one responsibility?
Yes — clearly. A single flow includes:

1. Input validation and conversion to UUID. 
2. Reading the main challenge.
3. Extracting filtering criteria.
4. Searching and filtering candidate challenges.
5. Converting document → DTO.
6. Random selection and limiting to 3.
7. Constructing GenericResultDto.


This violates the Single Responsibility Principle and complicates unit testing.
How can this be avoided?

**Review by ivilarop — CHANGES_REQUESTED** [review]

File: itachallenge-challenge/src/main/java/com/itachallenge/challenge/service/ChallengeServiceImpl.java (hunk 5 above)
Comment (ivilarop) [reply itachallenge-challenge/src/main/java/com/itachallenge/challenge/service/ChallengeServiceImpl.java]: Thanks for the clarification.
It's great that it now returns 204 when there are no related challenges; semantically, it makes sense.

Since the other lists are currently still responding with 200 and an empty list, let's note this so we can align all the endpoints later:

Add a technical debt card:
"Define and apply a single criterion (200 + empty list vs. 204) to all lists and notify the frontend."

The description indicates that this endpoint is the first to use 204, so the front-end team can take it into account.

With that, we won't block the delivery, and the team can decide on the final pattern in the next iteration. Good work!

File: itachallenge-challenge/src/main/java/com/itachallenge/challenge/service/ChallengeServiceImpl.java (hunk 4 above)
Comment (isaac-diez) [reply itachallenge-challenge/src/main/java/com/itachallenge/challenge/service/ChallengeServiceImpl.java:122]: The task/issue in Taiga for the refactoring of getChallengesByFilter() is created:
[https://tree.taiga.io/project/luisvi-ita-challenges/issue/579](https://tree.taiga.io/project/luisvi-ita-challenges/issue/579)

Yes, it is very coupled. One solution could be to detach the auxiliary processes from the main method:

**Main method**
- getRelatedChallenges()

**Auxiliary methods**
- findChallengeById()
- findRelatedChallenges() applying buildFilterPredicate()
- buildResultDtobuildResultDto()

File: itachallenge-challenge/src/main/java/com/itachallenge/challenge/service/ChallengeServiceImpl.java (hunk 5 above)
Comment (isaac-diez) [reply itachallenge-challenge/src/main/java/com/itachallenge/challenge/service/ChallengeServiceImpl.java]: Thanks for the feedback.

The task is created as an issue in Taiga: [https://tree.taiga.io/project/luisvi-ita-challenges/issue/580](https://tree.taiga.io/project/luisvi-ita-challenges/issue/580)

File: itachallenge-challenge/src/main/java/com/itachallenge/challenge/service/ChallengeServiceImpl.java (hunk 4 above)
Comment (ivilarop) [reply itachallenge-challenge/src/main/java/com/itachallenge/challenge/service/ChallengeServiceImpl.java:122]: The idea was to have done this refactoring in this PR, but given the limited time available, the improvement of refactoring the getRelatedChallenges method is documented in the technical debt: https://tree.taiga.io/project/luisvi-ita-challenges/issue/579

The main objective of the technical debt must be clear: to analyze and standardize, as far as possible, the functionalities of the filter search and the search for associated challenges.

**Review by ivilarop — CHANGES_REQUESTED** [review]

File: itachallenge-challenge/src/main/java/com/itachallenge/challenge/service/ChallengeServiceImpl.java (hunk 4 above)
Comment (isaac-diez) [reply itachallenge-challenge/src/main/java/com/itachallenge/challenge/service/ChallengeServiceImpl.java:122]: OK, thanks for the comment. I've added your input to the issue [#579](https://tree.taiga.io/project/luisvi-ita-challenges/issue/579) in Taiga and went ahead and added a new section in the PR summary to document the technical debts left to resolve in future PRs.

Comment (ivilarop) [reply itachallenge-challenge/src/main/java/com/itachallenge/challenge/service/ChallengeServiceImpl.java:122]: ok thks!

**Review by ivilarop — APPROVED** [review]
